### ISSUE TITLE
JM A100 Hybrid Neural 실험

### ISSUE THREAD

**Issue by Jungminii-1114** [calico-cat17/LG-Aimers-9th#8]
## 1. 핵심 결론

```text
GPU 학습 및 패키징: 성공
2024 strict OOF 로컬 검증: 매우 큰 개선
공식 Dacon 실행: 최종 성공
공식 Public BSS: 1109.9797870635
당시 공식 최고 1150.3783882339 대비: -40.3986011704
최종 판정: OFFICIAL_NO_GO
```

A100에서 학습한 Hybrid Neural은 2024년 월내 walk-forward OOF에서는 강한 개선을 보였지만, 공식 평가에서는 반대 결과가 나왔다. 따라서 이 실험에서 가장 중요한 교훈은 **A100 모델의 학습 성공이나 큰 로컬 ΔBSS가 공식 점수 전이를 보장하지 않는다**는 점이다.

현재 목표인 Public BSS 1175까지는 공식 최고점에서 `+24.6216117661`이 더 필요하다. 아래 A100 계열 결과만으로 이 목표 달성확률을 85% 이상이라고 주장할 근거는 없다.

## 2. 실행 환경과 안전 범위

주요 환경은 다음과 같다.

| 항목 | 값 |
|---|---|
| GPU | NVIDIA A100-PCIE-40GB |
| GPU 메모리 | 39.39 GiB |
| CUDA driver 표시 버전 | 12.2 |
| 실험 전용 환경 | 실험 폴더 내부 `.venv` |
| Python | 3.11.15 |
| PyTorch | 2.5.1+cu121 |
| NumPy | 2.4.6 |
| pandas | 2.3.3 |
| scikit-learn | 1.8.0 |
| CatBoost | 1.2.10 |
| XGBoost | 2.0.3 |
| LightGBM | 4.3.0 |

## 3. 입력 모델과 기본 예측

Hybrid Neural은 완전히 새로운 단독 분류기가 아니라, 이미 존재하는 두 팀 모델의 예측 위에 학습한 residual correction이다.

- JM 입력: `submit_JM1_RobustTensorEB_pitcher_floor025.zip`
  - 공식 Public BSS: `1148.77306`
- JY 입력: `sub_JY_team_residual_scale015.zip`
  - 공식 Public BSS: `1130.3604943627`
- 학습 OOF: `JM_JY_Strict_OOF_Stacker_Results.zip`
  - 보유 fold: 2024년 253,507행
- 배포 anchor: JM/JY 공식 quadratic anti-blend
  - 공식 Public BSS: `1150.3783882339`

최종 신경망은 `game_type == "R"`인 행의 logit에만 보정을 적용하고 F행은
기본 예측을 유지하도록 설계했다.

```text
R: logit(p_final) = logit(p_base) + scale × neural_correction
F: p_final = p_base
```

## 4. 모델 설계

입력 surface는 JM/JY 예측 자체와 두 예측의 차이, 확신도, 경기 상황 및
time-safe `asof_*` 통계로 구성했다.

- 수치형 38개
  - `p_jm`, `p_jy`, 두 예측 차이와 절댓값
  - 점수차, 주자, LI, 승리 기대값
  - 투수·타자 누적 성공률과 최근 경기 통계
  - 투수 pitch-mix 통계
- 범주형 14개
  - 투수·타자 ID와 팀 ID
  - 좌우완, 이닝, 카운트, 아웃, 주자 상태, 월·요일 등

하나의 공유 encoder 위에서 다음 세 head를 동시에 학습했다.

1. `DCN head`: 명시적 cross block과 deep MLP
2. `Transformer head`: 수치 token과 범주 embedding token의 self-attention
3. `Dual head`: 투수·타자 embedding의 곱과 절댓값 차이를 이용한 상호작용

학습 손실은 각 head와 세 head 평균 예측의 Brier loss를 함께 줄이도록 구성했다.
최종 full refit은 seed `17`, `42`, `777` 세 개로 수행했다.

## 5. 로컬 검증 프로토콜

2024 R행에서 월 단위 strict walk-forward를 사용했다.

```text
과거 월만 학습 → 다음 월 OOS 예측
5~6월: variant와 scale 선택
7~9월: 선택 후 동결 평가
투수 단위 cluster bootstrap: 5,000회
```

비교한 correction variant는 `dcn`, `transformer`, `dual`, 세 head 평균과
DCN/Transformer 평균이며, scale은 `0.05~1.00` 범위에서 검사했다. 선택 결과는
`transformer × 1.00`이었다.

### 6.1 Smoke test

- import 성공
- CUDA 인식 성공
- GPU: A100-PCIE-40GB
- 2,048개 학습 표본과 512개 forward 출력 성공
- JM/JY 중첩 제출 패키지 실행 성공
- 최종 출력: `SMOKE_OK`, return code `0`

### 6.2 본 실험

| 항목 | 결과 |
|---|---:|
| 실제 실행 장치 | CUDA |
| 전체 suite 실행 시간 | 172.72초 |
| 선택 variant | Transformer |
| 선택 scale | 1.00 |
| 5~6월 selection ΔBSS | +73.8890 |
| selection 최악 월 | +13.4886 |
| 7~9월 frozen ΔBSS | +275.6561 |
| frozen 최악 월 | +182.4556 |
| frozen 최악 좌·우완 그룹 | +187.1287 |
| cluster 95% CI | `[+198.6645, +359.5571]` |
| bootstrap P(ΔBSS > 0) | 100% |
| 당시 로컬 판정 | `GO_STRICT` |

당시 보고서의 `projected_public_bss_shape_only=1426.0344`는 OOF ΔBSS를 공식
점수에 단순 합산한 **shape-only 참고치**였다. 실제 공식 점수 예측치나 보장값으로
사용할 수 없으며, 공식 결과가 이를 명확히 반증했다.

생성된 원본 패키지:

```text
submit_JM_HybridNeural_transformer_s1000.zip
SHA-256: 3d701cc0bd9a021d8ca39682fd79318deff615bf26278c5ba329eb0a455652b4
```

## 7. Scale 축소 실험

공식 제출 전 보정 강도를 줄인 `s0350`, `s0500`을 추가 생성했다. 이 단계는
새 GPU 학습이 아니라 이미 학습한 neural direction의 배포 scale을 줄인 것이다.

| Scale | 전체 OOS ΔBSS | 5~6월 | 7~9월 | 최악 월 | 로컬 판정 |
|---:|---:|---:|---:|---:|---|
| 0.35 | +83.7411 | +34.6958 | +122.9836 | +5.2475 | GO_CANDIDATE |
| 0.50 | +113.5152 | +46.6558 | +167.0112 | +7.3230 | GO_CANDIDATE |
| 1.00 | 미기록(동일 원본 방향) | +73.8890 | +275.6561 | +13.4886 | GO_STRICT |

제출 횟수와 과보정 위험을 고려해 `s0500`을 우선 공식 probe로 선택했다.

### 8.1 로컬/서버 parity

- 최종 결과 tar.gz를 Mac으로 회수
- 서버와 Mac SHA-256 일치 확인
- tar 무결성 검사 return code `0`
- 별도 deployment parity artifact도 체크섬과 tar 검사를 통과

### 8.2 발견된 호환성 문제

1. parity 도구 실행 중 XGBoost 2.0.3이 NumPy 2.x에서 제거된 `np.NaN`을
   참조하는 문제가 발생했다. 이는 검증 환경 호환성 문제였다.
2. 첫 Dacon 패키지는 중첩 base package root 처리 오류로 실행에 실패했다.
3. 다음 패키지는 pandas Copy-on-Write의 읽기 전용 NumPy view에 값을 쓰면서
   `assignment destination is read-only` 오류가 발생했다.

최종 COW v2에서는 주요 `to_numpy()`에 `copy=True`를 사용하고, numeric
encoder에 writable copy를 명시했다. 일반 read-only smoke와 결측치 강제
read-only smoke를 모두 통과했다. 모델 가중치와 data artifact는 원본과
bit-identical이며 추론 로직 자체는 변경하지 않았다.

최종 정상 실행 패키지:

```text
JM_HybridNeural_s0500_V2.zip
동일 최종본 SHA-256:
6d63da042c6b7d347979bf03cdd167af173f8332c0a55434f4daa85d6ddc7fcc
```

## 9. 공식 제출 결과

| 제출 | 공식 Public BSS | 최고점 대비 | 판정 |
|---|---:|---:|---|
| JM/JY quadratic anti-blend base | 1150.3783882339 | 기준 | 유지 |
| Hybrid Neural s0500 COW v2 | 1109.9797870635 | -40.3986011704 | OFFICIAL_NO_GO |

패키지는 공식 환경에서 정상 실행됐으므로 낮은 점수는 단순 설치 오류가 아니라
예측 방향의 공식 일반화 실패로 판단한다. 사전에 정한 규칙에 따라 `s1000` 공식
probe는 제출하지 않았고, 원본 neural direction은 폐기했다.

## 10. 후속 A100 작업공간 실험

아래 실험은 같은 A100 전용 작업공간의 artifact를 이용했지만, 대부분 재학습이
아닌 CPU 기반 OOF 분석·router·패키징 실험이다.

### 10.1 Neural head reliability 재감사

- 선택: 5~6월
- 확인: 7월
- 완전 동결: 8~9월
- 다시 선택된 정책: 원본 `transformer × 1.00`
- 7월 ΔBSS: `+182.4556`
- 8~9월 ΔBSS: `+317.3749`
- cluster 95% CI: `[+221.3993, +422.9671]`
- 판정: 로컬 `GO_THREE_STAGE`, 공식 확률은 `NOT_ESTIMABLE`

공식 실패 이후에는 이 결과도 2024 OOF에 특화된 로컬 신호로 재해석했다.

### 10.2 Pitcher-hand negative router

`pitcher_hand == 1`이고 Transformer correction이 음수이면서 다른 head와
부호가 합의될 때만 correction을 유지했다.

- strength 선택은 5~6월만 사용
- 7~9월 frozen에서는 개선
- 그러나 5~6월 selection bootstrap P(Δ>0)가 약 60%
- 6월 ΔBSS가 소폭 음수
- 판정: `NO_GO_SELECTION`

### 10.3 Season-maturity hand router

`R & pitcher_hand == 1 & game_month >= 8`에만 negative consensus correction을
적용했다.

| 버전 | 전체 ΔBSS | 95% CI | 비고 | 판정 |
|---|---:|---:|---|---|
| strength 1.00 | +5.1571 | `[+1.6215,+9.1547]` | hash cohort 1 CI 하한 음수 | NO_GO |
| strength 0.50 | +2.7262 | `[+0.9380,+4.7555]` | 두 hash cohort CI 하한 양수 | GO_EXPLORATORY_ROBUST |

두 번째 버전도 frozen 결과를 본 뒤 만든 post-hoc 가설이므로 공식 제출 후보로
승격하지 않았다.

### 10.4 Negative router confirmatory

고정한 negative-only router를 selection/frozen으로 다시 확인했다.

- selection ΔBSS: `+0.1026`
- selection 95% CI: `[-0.8187,+1.0402]`
- frozen ΔBSS: `+7.9307`
- frozen 95% CI: `[-1.8708,+18.6189]`
- 판정: `NO_GO`

후기 시즌 점추정치는 양수였지만 CI 하한이 모두 0 아래라 제출하지 않았다.

### 10.5 Strict OOS Affine-R calibration

Hybrid s0500 base 확률을 5~6월 R행에서 affine shrinkage한 후 7~9월에 동결
평가했다.

```text
intercept = 0.0792800073
slope     = 0.8100198014
strength  = 0.5
```

초기 5~9월 감사에서는 all ΔBSS `+214.56`, frozen ΔBSS `+274.99`로
`GO_LOCAL_STRICT`가 나왔다. 그러나 전체 2024 strict OOF 3~10월을 다시
검사하자 3월 `-57.66`, 10월 `-6.21`, 최악 월내 구간 `-523.93`이 확인됐다.

최종 판정은 `HOLD / NO_GO_AS_IS`이며 공식 제출하지 않는다.

## 11. 교차연도 A100 실험 상태

2022→2023→2024 strict transfer를 검증하는 후속 코드는 설계했다.

- FT-Transformer residual
- DCN residual
- 투수·타자 Dual interaction residual
- 세 head 평균과 sign-consensus

그러나 정확한 다음 입력이 확보되지 않았다.

```text
JY_15Member_Strict_OOF_Results.zip
  ├── jy_15member_strict_oof_2022.npz
  ├── jy_15member_strict_oof_2023.npz
  └── jy_15member_strict_oof_2024.npz
```

따라서 Cross-Year Hybrid Neural Transfer는 **설계 완료 / 실행 미완료**다.
2024 단일 연도의 큰 OOF 개선을 교차연도 증거로 표현해서는 안 된다.

## 12. 실험에서 얻은 교훈

1. A100은 학습 시간을 크게 단축했다. 본 suite는 약 2분 53초에 끝났다.
2. 계산 자원이 늘어도 strict validation의 범위 부족은 해결되지 않는다.
3. 월내 walk-forward와 cluster bootstrap을 통과해도 연도 이동과 공식 평가에서
   실패할 수 있다.
4. 공식 점수와 연결되지 않은 `projected_public_bss_shape_only`는 점수 예측으로
   사용하면 안 된다.
5. 작은 smoke test만으로는 전체 테스트의 결측치·Copy-on-Write 경로를 모두
   검사할 수 없다. 강제 결측치 및 read-only smoke가 필요하다.
6. 후속 router를 같은 2024 OOF에서 반복 선택하면 post-hoc 편향이 누적된다.
7. 다음 신경망 실험은 정확한 2022·2023 strict OOF를 먼저 확보한 뒤에만 진행해야
   한다.
